File: Javascript/Stack/M-735. Asteroid Collision.ts
function asteroidCollision(asteroids: number[]): number[] {
  const stack: number[] = [];
  for (let i = 0; i < asteroids.length; i++) {
    if (!stack.length) {
      stack.push(asteroids[i]);
      continue;
    }
    /** no explode happened:
     * 1. lastAsteroid and asteroids[i] goes same direction or
     * 2. lastAsteroid goes left and asteroids[i] gose right
     **/
    const lastAsteroid = stack[stack.length - 1];
    if (
      lastAsteroid * asteroids[i] > 0 ||
      (lastAsteroid < 0 && asteroids[i] > 0)
    ) {
      stack.push(asteroids[i]);
      continue;
    }
    /** explode happened */
    if (Math.abs(lastAsteroid) > Math.abs(asteroids[i])) continue;
    stack.pop();
    if (Math.abs(lastAsteroid) < Math.abs(asteroids[i])) i--;
  }
  //console.log('stack', stack);
  return stack;
}

const testData = [
  {
    asteroids: [5, 10, -5],
    ans: [5, 10],
  },
  {
    asteroids: [8, -8],
    ans: [],
  },
  {
    asteroids: [10, 2, -5],
    ans: [10],
  },
  { asteroids: [-2, -1, 1, 2], ans: [-2, -1, 1, 2] },
];

for (const data of testData) {
  console.log(
    JSON.stringify(data['ans']) ===
      JSON.stringify(asteroidCollision(data['asteroids']))
  );
}

File: Stack/M-735. Asteroid Collision/M-735. Asteroid Collision.py
class Solution:
    @classmethod
    def asteroidCollision(self, asteroids: list[int]) -> list[int]:
        stack = []
        i = 0
        while i < len(asteroids):
            if not len(stack):
                stack.append(asteroids[i])
                i += 1
                continue

            # no explode happened,
            # same direction or last goes left and current goes right
            last_asteroid = stack[len(stack) - 1]
            if last_asteroid * asteroids[i] > 0 or (
                last_asteroid < 0 and asteroids[i] > 0
            ):
                stack.append(asteroids[i])
                i += 1
                continue

            # explode happened
            if abs(last_asteroid) > abs(asteroids[i]):
                i += 1
                continue
            stack.pop()
            if abs(last_asteroid) == abs(asteroids[i]):
                i += 1
        return stack


testData = [
    {
        "asteroids": [5, 10, -5],
        "ans": [5, 10],
    },
    {
        "asteroids": [8, -8],
        "ans": [],
    },
    {
        "asteroids": [10, 2, -5],
        "ans": [10],
    },
    {"asteroids": [-2, -1, 1, 2], "ans": [-2, -1, 1, 2]},
]

for data in testData:
    print(data["ans"] == Solution.asteroidCollision(data["asteroids"]))
